swupd/manifest.go

```diff
@@ -755,22 +756,60 @@ func (m *Manifest) addManifestFiles(ui UpdateInfo, c config) error {
 		}
 	} else {
 		// Add files to manifest that do not exist in included bundles.
+		chrootDir := filepath.Join(c.imageBase, fmt.Sprint(ui.version), "full")
 		includes := m.GetRecursiveIncludes()
+		usedDirs := make(map[string]struct{})
+		bundleDirs := []string{}
+		bundleFiles := []string{}
 		for f := range m.BundleInfo.Files {
+			fullPath := filepath.Join(chrootDir, f)
+			if fi, err := os.Lstat(fullPath); err == nil {
+				if !fi.IsDir() {
+					usedDirs[filepath.Dir(f)] = struct{}{}
+					bundleFiles = append(bundleFiles, f)
+				} else {
+					bundleDirs = append(bundleDirs, f)
+				}
+			}
+		}
+		for _, f := range bundleFiles {
 			isIncluded := false
 			for _, inc := range includes {
 				// Handle cycles
 				if inc.Name == m.Name {
 					continue
 				}
-				chrootDir := filepath.Join(c.imageBase, fmt.Sprint(ui.version), "full")
-				fullPath := filepath.Join(chrootDir, f)
-				if fi, err := os.Lstat(fullPath); err == nil {
-					if !fi.IsDir() {
-						if _, ok := inc.BundleInfo.Files[f]; ok {
-							isIncluded = true
-							break
-						}
+				if _, ok := inc.BundleInfo.Files[f]; ok {
+					isIncluded = true
+					break
+				}
+			}
+			if !isIncluded {
+				if err := m.addFile(f, c, ui.version); err != nil {
+					return err
+				}
+			}
+		}
+		// Directories are sorted into reverse order to allow usedDirs
+		// to populate correctly (/a/b/c will be seen before /a/b which
+		// will allow /a to be set when /a/b is seen)
+		slices.Sort(bundleDirs)
+		slices.Reverse(bundleDirs)
+		for _, f := range bundleDirs {
+			if _, ok := usedDirs[f]; ok {
+				usedDirs[filepath.Dir(f)] = struct{}{}
+			}
+		}
+		for _, f := range bundleDirs {
+			isIncluded := false
+			for _, inc := range includes {
```

Comment:
Do we need to fudge `includes` so `os-core` is always the first element?

Reply:
In Manifest files I believe os-core is already the first element but mixer shouldn't be impacted by include order. The subtraction should happen regardless if something is in os-core no matter if os-core is looked at first or last from that logic.